python/ray/_private/node.py

```diff
@@ -577,23 +571,53 @@ def is_head(self):
 
     def get_gcs_client(self):
         if self._gcs_client is None:
-            for _ in range(NUM_REDIS_GET_RETRIES):
-                gcs_address = None
-                last_ex = None
-                try:
-                    gcs_address = self.gcs_address
-                    self._gcs_client = GcsClient(address=gcs_address)
+            self._init_gcs_client()
+        return self._gcs_client
+
+    def _init_gcs_client(self):
+        if self.head:
+            gcs_process = self.all_processes[ray_constants.PROCESS_TYPE_GCS_SERVER][
+                0
+            ].process
+        else:
+            gcs_process = None
+
+        for _ in range(ray_constants.NUM_REDIS_GET_RETRIES):
+            gcs_address = None
+            last_ex = None
+            try:
+                gcs_address = self.gcs_address
+                client = GcsClient(address=gcs_address)
+                if self.head:
+                    # Send a simple request to make sure GCS is alive
+                    # if it's a head node.
+                    client.internal_kv_get(b"dummy", None)
+                self._gcs_client = client
+                break
+            except Exception:
+                if gcs_process is not None and gcs_process.poll() is not None:
+                    # GCS has exited.
                     break
-                except Exception:
-                    last_ex = traceback.format_exc()
-                    logger.debug(f"Connecting to GCS: {last_ex}")
-                    time.sleep(1)
-            assert self._gcs_client is not None, (
-                f"Failed to connect to GCS at address={gcs_address}. "
-                f"Last exception: {last_ex}"
+                last_ex = traceback.format_exc()
+                logger.debug(f"Connecting to GCS: {last_ex}")
+                time.sleep(1)
+
+        if self._gcs_client is None:
+            with open(os.path.join(self._logs_dir, "gcs_server.err")) as err:
+                # Use " C " or " E " to exclude the stacktrace.
+                # This should work for most cases, especitally
+                # it's when GCS is starting. Only display last 10 lines of logs.
+                errors = [e for e in err.readlines() if " C " in e or " E " in e][-10:]
+            error_msg = "\n" + "".join(errors) + "\n"
+            logger.fatal(
+                f"Failed to start GCS. Last {len(errors)} lines of error files:"
+                f"{error_msg}"
+                f"Please check {os.path.join(self._logs_dir, 'gcs_server.out')}"
+                " for details"
```

Comment:
does it make sense to include this full message in the exception stacktrace raised below? That way the console output will be easier to read. Right now it is:

```
<detailed error message about GCS failure>
<very long stack trace>
<very long stack trace>
...
<very long stack trace>
Exception: "Failed to start GCS."
```

but if you make this the exception message, it would be:
```
<very long stack trace>
<very long stack trace>
...
<very long stack trace>
Exception: "Failed to start GCS. Last 10 lines of error files:"
<detailed error message about GCS failure>
```

that should be easier to read, puts the most important info right at the bottom of the stack trace

Reply:
sgtm. let me update this and post the error message again.